File: src/core/providers/anthropic.ts
/**
 * Anthropic Provider Implementation
 *
 * Provides integration with Anthropic's API for completions and streaming.
 */

import Anthropic from '@anthropic-ai/sdk';
import type {
  Message,
  CompletionOptions,
  CompletionResult,
  StreamChunk,
  MessageContent,
  TextContent,
} from '../../types';
import { BaseProvider, BaseProviderConfig } from './base';
import { Logger } from '../../utils/logger';

/**
 * Anthropic provider configuration
 */
export interface AnthropicConfig extends BaseProviderConfig {
  apiKey: string;
  baseUrl?: string;
  defaultModel?: string;
  timeout?: number;
  version?: string;
}

/**
 * Anthropic Provider class
 *
 * Implements AI provider interface using Anthropic's API.
 */
export class AnthropicProvider extends BaseProvider {
  private client: Anthropic;
  private version: string;

  constructor(config: AnthropicConfig, logger?: Logger) {
    super(config, logger);
    this.version = config.version ?? '2023-06-01';

    this.validateApiKey();

    // Initialize Anthropic client
    this.client = new Anthropic({
      apiKey: config.apiKey,
      baseURL: config.baseUrl,
      timeout: config.timeout ?? 30000,
      maxRetries: 3,
      defaultHeaders: {
        'anthropic-version': this.version,
      },
    });

    this.logger.info('Anthropic provider initialized', 'AnthropicProvider', {
      baseUrl: config.baseUrl,
      defaultModel: config.defaultModel,
      version: this.version,
    });
  }

  /**
   * Get provider name
   */
  get name(): string {
    return 'Anthropic';
  }

  /**
   * Get provider prefix
   */
  get prefix(): 'anthropic' {
    return 'anthropic';
  }

  /**
   * Initialize provider capabilities
   */
  protected initializeCapabilities() {
    return {
      streaming: true,
      tools: true,
      vision: true,
      maxTokens: 200000,
    };
  }

  /**
   * Generate a complete completion
   *
   * @param model - Model identifier
   * @param messages - Array of messages
   * @param options - Completion options
   * @returns Completion result
   */
  async complete(
    model: string,
    messages: Message[],
    options?: CompletionOptions
  ): Promise<CompletionResult> {
    this.validateMessages(messages);
    this.validateOptions(options);

    this.logger.debug('Starting Anthropic completion', 'AnthropicProvider', {
      model,
      messageCount: messages.length,
      options,
    });

    try {
      const { system, anthropicMessages } = this.convertMessages(messages);

      const response = await this.client.messages.create({
        model,
        messages: anthropicMessages,
        system,
        max_tokens: options?.maxTokens ?? 4096,
        temperature: options?.temperature,
        top_p: options?.topP,
        tools: this.convertTools(options?.tools),
        stream: false,
      });

      const content = this.parseContent(response.content);

      const result: CompletionResult = {
        content,
        model: response.model,
        stopReason: response.stop_reason,
        usage: response.usage
          ? {
              inputTokens: response.usage.input_tokens,
              outputTokens: response.usage.output_tokens,
              totalTokens: response.usage.input_tokens + response.usage.output_tokens,
            }
          : undefined,
      };

      this.logger.debug('Anthropic completion successful', 'AnthropicProvider', {
        model: result.model,
        stopReason: result.stopReason,
        usage: result.usage,
      });

      return result;
    } catch (error) {
      this.handleError(error);
    }
  }

  /**
   * Generate a streaming completion
   *
   * @param model - Model identifier
   * @param messages - Array of messages
   * @param options - Completion options
   * @returns Async generator of stream chunks
   */
  async *stream(
    model: string,
    messages: Message[],
    options?: CompletionOptions
  ): AsyncGenerator<StreamChunk> {
    this.validateMessages(messages);
    this.validateOptions(options);

    this.logger.debug('Starting Anthropic streaming', 'AnthropicProvider', {
      model,
      messageCount: messages.length,
      options,
    });

    try {
      const { system, anthropicMessages } = this.convertMessages(messages);

      const stream = await this.client.messages.create({
        model,
        messages: anthropicMessages,
        system,
        max_tokens: options?.maxTokens ?? 4096,
        temperature: options?.temperature,
        top_p: options?.topP,
        tools: this.convertTools(options?.tools),
        stream: true,
      });

      let fullContent = '';

      for await (const event of stream) {
        if (event.type === 'content_block_delta') {
          if (event.delta.type === 'text_delta' && event.delta.text) {
            fullContent += event.delta.text;

            const streamChunk: StreamChunk = {
              content: { type: 'text', text: event.delta.text } as TextContent,
              delta: event.delta.text,
              done: false,
            };

            yield streamChunk;
          }
        } else if (event.type === 'message_stop') {
          const finalChunk: StreamChunk = {
            content: { type: 'text', text: fullContent } as TextContent,
            delta: '',
            done: true,
            usage: event.message?.usage
              ? {
                  inputTokens: event.message.usage.input_tokens,
                  outputTokens: event.message.usage.output_tokens,
                  totalTokens: event.message.usage.input_tokens + event.message.usage.output_tokens,
                }
              : undefined,
          };

          yield finalChunk;

          this.logger.debug('Anthropic streaming complete', 'AnthropicProvider', {
            model,
            stopReason: event.message?.stop_reason,
            usage: finalChunk.usage,
          });

          break;
        }
      }
    } catch (error) {
      this.handleError(error);
    }
  }

  /**
   * Count tokens in messages
   *
   * @param messages - Array of messages
   * @returns Token count
   */
  async countTokens(messages: Message[]): Promise<number> {
    // Anthropic provides a token counting method
    const { anthropicMessages, system } = this.convertMessages(messages);

    try {
      const count = await this.client.messages.countTokens({
        model: this.getDefaultModel(),
        messages: anthropicMessages,
        system,
      });

      this.logger.debug('Token count', 'AnthropicProvider', {
        messageCount: messages.length,
        tokens: count.input_tokens,
      });

      return count.input_tokens;
    } catch (error) {
      // Fallback to approximation if token counting fails
      this.logger.warn('Token counting failed, using approximation', 'AnthropicProvider', { error });
      const text = this.extractTextFromMessages(messages);
      return Math.ceil(text.length / 4);
    }
  }

  /**
   * Convert messages to Anthropic format
   *
   * @param messages - Array of messages
   * @returns Anthropic message format with system prompt
   */
  private convertMessages(messages: Message[]): {
    system: string | undefined;
    anthropicMessages: Anthropic.MessageParam[];
  } {
    const systemMessages: string[] = [];
    const anthropicMessages: Anthropic.MessageParam[] = [];

    for (const msg of messages) {
      if (msg.role === 'system') {
        const text = this.extractText(msg.content);
        if (text) {
          systemMessages.push(text);
        }
      } else {
        anthropicMessages.push({
          role: msg.role as 'user' | 'assistant',
          content: this.convertContent(msg.content),
        });
      }
    }

    return {
      system: systemMessages.length > 0 ? systemMessages.join('\n\n') : undefined,
      anthropicMessages,
    };
  }

  /**
   * Convert content to Anthropic format
   *
   * @param content - Message content
   * @returns Anthropic content format
   */
  private convertContent(content: Message['content']): Anthropic.ContentBlock[] {
    if (typeof content === 'string') {
      return [{ type: 'text', text: content }];
    }

    if (Array.isArray(content)) {
      return content.map((item) => {
        if (typeof item === 'string') {
          return { type: 'text', text: item };
        }
        if (item.type === 'text') {
          return { type: 'text', text: item.text };
        }
        if (item.type === 'image') {
          return {
            type: 'image',
            source: {
              type: item.source.type === 'url' ? 'url' : 'base64',
              [item.source.type === 'url' ? 'url' : 'media_type']: item.source.type === 'url'
                ? item.source.data
                : item.source.media_type || 'image/png',
              data: item.source.type === 'url' ? undefined : item.source.data,
            },
          };
        }
        if (item.type === 'tool_use') {
          return {
            type: 'tool_use',
            id: item.id,
            name: item.name,
            input: item.input,
          };
        }
        if (item.type === 'tool_result') {
          return {
            type: 'tool_result',
            tool_use_id: item.tool_use_id,
            content: this.convertContent(item.content ?? ''),
            is_error: item.is_error,
          };
        }
        return { type: 'text', text: '' };
      });
    }

    if (typeof content === 'object' && content !== null) {
      if (content.type === 'text') {
        return [{ type: 'text', text: content.text }];
      }
      if (content.type === 'image') {
        return [
          {
            type: 'image',
            source: {
              type: content.source.type === 'url' ? 'url' : 'base64',
              [content.source.type === 'url' ? 'url' : 'media_type']: content.source.type === 'url'
                ? content.source.data
                : content.source.media_type || 'image/png',
              data: content.source.type === 'url' ? undefined : content.source.data,
            },
          },
        ];
      }
      if (content.type === 'tool_use') {
        return [
          {
            type: 'tool_use',
            id: content.id,
            name: content.name,
            input: content.input,
          },
        ];
      }
      if (content.type === 'tool_result') {
        return [
          {
            type: 'tool_result',
            tool_use_id: content.tool_use_id,
            content: this.convertContent(content.content ?? ''),
            is_error: content.is_error,
          },
        ];
      }
    }

    return [{ type: 'text', text: '' }];
  }

  /**
   * Parse content from Anthropic response
   *
   * @param content - Content from Anthropic
   * @returns Parsed message content
   */
  private parseContent(content: Anthropic.ContentBlock[]): MessageContent {
    if (content.length === 0) {
      return { type: 'text', text: '' };
    }

    if (content.length === 1) {
      const block = content[0];

      if (block.type === 'text') {
        return { type: 'text', text: block.text };
      }

      return block as MessageContent;
    }

    // Multiple content blocks
    return content as MessageContent[];
  }

  /**
   * Extract text from all messages
   *
   * @param messages - Array of messages
   * @returns Combined text
   */
  private extractTextFromMessages(messages: Message[]): string {
    return messages.map((msg) => this.extractText(msg.content)).join('');
  }

  /**
   * Convert tools to Anthropic format
   *
   * @param tools - Array of tools
   * @returns Anthropic tool format
   */
  protected convertTools(tools?: any[]): Anthropic.Tool[] | undefined {
    if (!tools || tools.length === 0) {
      return undefined;
    }

    return tools.map((tool) => ({
      name: tool.name,
      description: tool.description,
      input_schema: tool.inputSchema,
    }));
  }
}

File: internal/llm/providers/anthropic.go
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
	"github.com/anthropics/anthropic-sdk-go/shared/constant"
	"github.com/komplete-kontrol/cli/internal/llmapi"
)

// AnthropicProvider implements Provider for Anthropic API
type AnthropicProvider struct {
	client  anthropic.Client
	apiKey  string
	baseURL string
	models  []llmapi.ModelInfo
}

// AnthropicConfig holds Anthropic provider configuration
type AnthropicConfig struct {
	APIKey  string
	BaseURL string
}

// NewAnthropicProvider creates a new Anthropic provider
func NewAnthropicProvider(cfg AnthropicConfig) *AnthropicProvider {
	opts := []option.RequestOption{option.WithAPIKey(cfg.APIKey)}
	if cfg.BaseURL != "" {
		opts = append(opts, option.WithBaseURL(cfg.BaseURL))
	}
	client := anthropic.NewClient(opts...)

	baseURL := cfg.BaseURL
	if baseURL == "" {
		baseURL = "https://api.anthropic.com"
	}

	return &AnthropicProvider{
		client:  client,
		apiKey:  cfg.APIKey,
		baseURL: baseURL,
		models:  getAnthropicModels(),
	}
}

// Name returns the provider name
func (p *AnthropicProvider) Name() string {
	return "Anthropic"
}

// Type returns the provider type
func (p *AnthropicProvider) Type() llmapi.ProviderType {
	return llmapi.ProviderAnthropic
}

// Complete sends a completion request
func (p *AnthropicProvider) Complete(ctx context.Context, req *llmapi.CompletionRequest) (*llmapi.CompletionResponse, error) {
	// Convert messages to Anthropic format (system prompt is top-level, not a role)
	var messages []anthropic.MessageParam
	for _, msg := range req.Messages {
		if msg.Role == "system" {
			continue
		}
		if msg.Role == "assistant" {
			messages = append(messages, anthropic.NewAssistantMessage(anthropic.NewTextBlock(msg.Content)))
		} else {
			messages = append(messages, anthropic.NewUserMessage(anthropic.NewTextBlock(msg.Content)))
		}
	}
	if len(messages) == 0 {
		// Ensure we always send at least one user message.
		messages = append(messages, anthropic.NewUserMessage(anthropic.NewTextBlock("")))
	}

	// Convert tools to Anthropic format
	var tools []anthropic.ToolUnionParam
	if len(req.Tools) > 0 {
		tools = make([]anthropic.ToolUnionParam, 0, len(req.Tools))
		for _, tool := range req.Tools {
			u := anthropic.ToolUnionParamOfTool(
				anthropic.ToolInputSchemaParam{
					Type:       constant.Object("object"),
					Properties: tool.InputSchema["properties"],
					Required:   asStringSlice(tool.InputSchema["required"]),
				},
				tool.Name,
			)
			if tool.Description != "" && u.OfTool != nil {
				u.OfTool.Description = anthropic.String(tool.Description)
			}
			tools = append(tools, u)
		}
	}

	// Create request
	anthropicReq := anthropic.MessageNewParams{
		Model:     anthropic.Model(req.Model),
		Messages:  messages,
		MaxTokens: int64(req.MaxTokens),
		Tools:     tools,
	}
	if req.System != "" {
		anthropicReq.System = []anthropic.TextBlockParam{{
			Type: constant.Text("text"),
			Text: req.System,
		}}
	}
	if req.Temperature != 0 {
		anthropicReq.Temperature = anthropic.Float(req.Temperature)
	}

	// Send request
	resp, err := p.client.Messages.New(ctx, anthropicReq)
	if err != nil {
		return nil, fmt.Errorf("anthropic request failed: %w", err)
	}

	// Convert response
	return p.convertResponse(resp), nil
}

// StreamComplete sends a streaming completion request
func (p *AnthropicProvider) StreamComplete(ctx context.Context, req *llmapi.CompletionRequest, handler llmapi.StreamHandler) (*llmapi.CompletionResponse, error) {
	// Convert messages to Anthropic format (system prompt is top-level, not a role)
	var messages []anthropic.MessageParam
	for _, msg := range req.Messages {
		if msg.Role == "system" {
			continue
		}
		if msg.Role == "assistant" {
			messages = append(messages, anthropic.NewAssistantMessage(anthropic.NewTextBlock(msg.Content)))
		} else {
			messages = append(messages, anthropic.NewUserMessage(anthropic.NewTextBlock(msg.Content)))
		}
	}
	if len(messages) == 0 {
		messages = append(messages, anthropic.NewUserMessage(anthropic.NewTextBlock("")))
	}

	// Convert tools to Anthropic format
	var tools []anthropic.ToolUnionParam
	if len(req.Tools) > 0 {
		tools = make([]anthropic.ToolUnionParam, 0, len(req.Tools))
		for _, tool := range req.Tools {
			u := anthropic.ToolUnionParamOfTool(
				anthropic.ToolInputSchemaParam{
					Type:       constant.Object("object"),
					Properties: tool.InputSchema["properties"],
					Required:   asStringSlice(tool.InputSchema["required"]),
				},
				tool.Name,
			)
			if tool.Description != "" && u.OfTool != nil {
				u.OfTool.Description = anthropic.String(tool.Description)
			}
			tools = append(tools, u)
		}
	}

	anthropicReq := anthropic.MessageNewParams{
		Model:     anthropic.Model(req.Model),
		Messages:  messages,
		MaxTokens: int64(req.MaxTokens),
		Tools:     tools,
	}
	if req.System != "" {
		anthropicReq.System = []anthropic.TextBlockParam{{
			Type: constant.Text("text"),
			Text: req.System,
		}}
	}
	if req.Temperature != 0 {
		anthropicReq.Temperature = anthropic.Float(req.Temperature)
	}

	// Send streaming request
	stream := p.client.Messages.NewStreaming(ctx, anthropicReq)

	var fullContent strings.Builder
	var toolCalls []llmapi.ToolCall

	var message anthropic.Message
	for stream.Next() {
		event := stream.Current()
		if err := message.Accumulate(event); err != nil {
			handler.OnError(fmt.Errorf("anthropic stream accumulate error: %w", err))
			return nil, err
		}

		if event.Type == "content_block_delta" {
			delta := event.AsContentBlockDelta()
			if delta.Delta.Type == "text_delta" && delta.Delta.Text != "" {
				handler.OnToken(delta.Delta.Text)
				fullContent.WriteString(delta.Delta.Text)
			}
		}

		if event.Type == "message_stop" {
			break
		}
	}
	if err := stream.Err(); err != nil {
		handler.OnError(fmt.Errorf("anthropic stream error: %w", err))
		return nil, err
	}

	// Extract tool calls from accumulated message
	for _, block := range message.Content {
		if block.Type == "tool_use" {
			args := map[string]any{}
			_ = json.Unmarshal(block.Input, &args)
			toolCalls = append(toolCalls, llmapi.ToolCall{ID: block.ID, Name: block.Name, Arguments: args})
		}
	}

	resp := &llmapi.CompletionResponse{
		ID:    message.ID,
		Model: string(message.Model),
		Message: llmapi.Message{
			Role:    "assistant",
			Content: fullContent.String(),
		},
		ToolCalls: toolCalls,
		TokensUsed: llmapi.TokenUsage{
			InputTokens:  int(message.Usage.InputTokens),
			OutputTokens: int(message.Usage.OutputTokens),
			TotalTokens:  int(message.Usage.InputTokens + message.Usage.OutputTokens),
		},
		StopReason: string(message.StopReason),
	}
	handler.OnDone(resp)
	return resp, nil
}

// ListModels returns available Anthropic models
func (p *AnthropicProvider) ListModels(ctx context.Context) ([]llmapi.ModelInfo, error) {
	return p.models, nil
}

// IsAvailable checks if the provider is available
func (p *AnthropicProvider) IsAvailable(ctx context.Context) bool {
	if p.apiKey == "" {
		return false
	}

	// Simple health check
	req, err := http.NewRequestWithContext(ctx, "GET", p.baseURL+"/v1/messages", nil)
	if err != nil {
		return false
	}
	req.Header.Set("x-api-key", p.apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	client := &http.Client{Timeout: 5 * 1000000000} // 5 seconds
	resp, err := client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusUnauthorized
}

// convertResponse converts Anthropic response to standard format
func (p *AnthropicProvider) convertResponse(resp *anthropic.Message) *llmapi.CompletionResponse {
	// Extract content
	var content string
	if len(resp.Content) > 0 {
		// Prefer first text block; tool_use blocks have no text.
		for _, block := range resp.Content {
			if block.Type == "text" {
				content = block.Text
				break
			}
		}
	}

	// Extract tool calls
	var toolCalls []llmapi.ToolCall
	for _, block := range resp.Content {
		if block.Type == "tool_use" {
			args := map[string]any{}
			_ = json.Unmarshal(block.Input, &args)
			toolCalls = append(toolCalls, llmapi.ToolCall{ID: block.ID, Name: block.Name, Arguments: args})
		}
	}

	return &llmapi.CompletionResponse{
		ID:    resp.ID,
		Model: string(resp.Model),
		Message: llmapi.Message{
			Role:    "assistant",
			Content: content,
		},
		ToolCalls: toolCalls,
		TokensUsed: llmapi.TokenUsage{
			InputTokens:  int(resp.Usage.InputTokens),
			OutputTokens: int(resp.Usage.OutputTokens),
			TotalTokens:  int(resp.Usage.InputTokens + resp.Usage.OutputTokens),
		},
		StopReason: string(resp.StopReason),
	}
}

// getAnthropicModels returns available Anthropic models
func getAnthropicModels() []llmapi.ModelInfo {
	return []llmapi.ModelInfo{
		{
			ID:       "claude-sonnet-4-5-20250514",
			Name:     "Claude Sonnet 4.5",
			Provider: llmapi.ProviderAnthropic,
			Capabilities: llmapi.ModelCapabilities{
				Streaming:  true,
				Tools:      true,
				Vision:     true,
				Multimodal: true,
			},
			CostPer1k: 0.003,
			MaxTokens: 200000,
		},
		{
			ID:       "claude-3-5-sonnet-20241022",
			Name:     "Claude 3.5 Sonnet",
			Provider: llmapi.ProviderAnthropic,
			Capabilities: llmapi.ModelCapabilities{
				Streaming:  true,
				Tools:      true,
				Vision:     true,
				Multimodal: true,
			},
			CostPer1k: 0.003,
			MaxTokens: 200000,
		},
		{
			ID:       "claude-3-5-haiku-20241022",
			Name:     "Claude 3.5 Haiku",
			Provider: llmapi.ProviderAnthropic,
			Capabilities: llmapi.ModelCapabilities{
				Streaming:  true,
				Tools:      true,
				Vision:     true,
				Multimodal: true,
			},
			CostPer1k: 0.00025,
			MaxTokens: 200000,
		},
		{
			ID:       "claude-3-haiku-20240307",
			Name:     "Claude 3 Haiku",
			Provider: llmapi.ProviderAnthropic,
			Capabilities: llmapi.ModelCapabilities{
				Streaming:  true,
				Tools:      true,
				Vision:     true,
				Multimodal: true,
			},
			CostPer1k: 0.00025,
			MaxTokens: 200000,
		},
	}
}

func asStringSlice(v any) []string {
	if v == nil {
		return nil
	}
	arr, ok := v.([]any)
	if !ok {
		// some callers may already provide []string
		if ss, ok := v.([]string); ok {
			return ss
		}
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, x := range arr {
		if s, ok := x.(string); ok {
			out = append(out, s)
		}
	}
	return out
}
